packages/builders/src/messages/embed/Embed.ts

```diff
@@ -1,86 +1,64 @@
-import type { APIEmbed, APIEmbedAuthor, APIEmbedField, APIEmbedFooter, APIEmbedImage } from 'discord-api-types/v10';
-import { normalizeArray, type RestOrArray } from '../../util/normalizeArray.js';
-import {
-	colorPredicate,
-	descriptionPredicate,
-	embedAuthorPredicate,
-	embedFieldsArrayPredicate,
-	embedFooterPredicate,
-	imageURLPredicate,
-	timestampPredicate,
-	titlePredicate,
-	urlPredicate,
-	validateFieldLength,
-} from './Assertions.js';
+import type { JSONEncodable } from '@discordjs/util';
+import type { APIEmbed, APIEmbedAuthor, APIEmbedField, APIEmbedFooter } from 'discord-api-types/v10';
+import type { RestOrArray } from '../../util/normalizeArray.js';
+import { normalizeArray } from '../../util/normalizeArray.js';
+import { isValidationEnabled } from '../../util/validation.js';
+import { embedPredicate } from './Assertions.js';
+import { EmbedAuthorBuilder } from './EmbedAuthor.js';
+import { EmbedFieldBuilder } from './EmbedField.js';
+import { EmbedFooterBuilder } from './EmbedFooter.js';
 
 /**
- * A tuple satisfying the RGB color model.
- *
- * @see {@link https://developer.mozilla.org/docs/Glossary/RGB}
+ * Data stored in the process of constructing an embed.
  */
-export type RGBTuple = [red: number, green: number, blue: number];
+export interface EmbedBuilderData extends Omit<APIEmbed, 'author' | 'fields' | 'footer'> {
+	author?: EmbedAuthorBuilder;
+	fields?: EmbedFieldBuilder[];
+	footer?: EmbedFooterBuilder;
+}
 
 /**
- * The base icon data typically used in payloads.
+ * A builder that creates API-compatible JSON data for embeds.
  */
-export interface IconData {
+export class EmbedBuilder implements JSONEncodable<APIEmbed> {
 	/**
-	 * The URL of the icon.
+	 * The API data associated with this embed.
 	 */
-	iconURL?: string;
+	private readonly data: EmbedBuilderData;
+
 	/**
-	 * The proxy URL of the icon.
+	 * Gets the author of this embed.
 	 */
-	proxyIconURL?: string;
-}
-
-/**
- * Represents the author data of an embed.
- */
-export interface EmbedAuthorData extends IconData, Omit<APIEmbedAuthor, 'icon_url' | 'proxy_icon_url'> {}
-
-/**
- * Represents the author options of an embed.
- */
-export interface EmbedAuthorOptions extends Omit<EmbedAuthorData, 'proxyIconURL'> {}
-
-/**
- * Represents the footer data of an embed.
- */
-export interface EmbedFooterData extends IconData, Omit<APIEmbedFooter, 'icon_url' | 'proxy_icon_url'> {}
-
-/**
- * Represents the footer options of an embed.
- */
-export interface EmbedFooterOptions extends Omit<EmbedFooterData, 'proxyIconURL'> {}
+	public get author(): EmbedAuthorBuilder {
+		return (this.data.author ??= new EmbedAuthorBuilder());
```

Comment:
One small issue: consider this code
```ts
const builder = new EmbedBuilder();

const author = builder.author; // data.author is set

if (condition) {
	author.setName('idk idc');
}

const json = builder.toJSON();
```

Won't this throw on Discord's side because author is an empty object? Ideally we'd ignore footer/author if objects are empty.

Yes, I'm aware you should do builder.author in that if check, but in the EVENT someone doesn't

Reply:
> Won't this throw on Discord's side because author is an empty object? 

It'll throw on our side, even, but doesn't seem ideal indeed.

> Ideally we'd ignore footer/author if objects are empty.

Not sure if this is the solution either 😅, now that you present this issue I was there was an easy way to prevent assigning in the getter only if the user actually did something with it. Could we maybe approach a different API for those?

Refactor the getter to `get author(): EmbedAuthorBuilder | undefined`
```ts
return this.data.author;
```

and provide
```ts
updateAuthor(updater: EmbedAuthorBuilder) => unknown): this;
```

Which would work essentially like `setAuthor`, except it only takes the callback and internally it just like,
```ts
this.data.author ??= new EmbedAuthorBuilder();
updater(this.data.author);

return this;
```

so you'd use it like
```ts
const builder = new EmbedBuilder(existingData);
// Let's imagine the user only wants to update it if it has even been sent, all though the update call
// would work just fine without this if:
if (builder.author) {
  builder.updateAuthor(author => author.setName('hello'));
}
```

Reply:
I see why we'd need an updateAuthor method. Also I don't get why we have getters for author/footer but nothing else 👀 

Imo
- drop getter
- set/updateAuthor

Reply:
> Also I don't get why we have getters for author/footer but nothing else 👀

The idea was that we're never exposing raw fields/primitives and instead via the getter you're able to `builder.author.setName()`, for instance, but yes, I'll be adopting the new pattern.